### PR TITLE
marcel: init at 0.22.2

### PR DESCRIPTION
**Pull request by KUD-00**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

It's my first time to contribute. I apologize first for any mistakes in the codes. I have tested the binary in /result/bin(in `nix-shell - A marcel` environment), but not sure if it is executable when installing it on any system. I tried `$ nix-env -f . -iA marcel` but it said it can't find `lib` and many other parameters. I'm not sure is it good to use `/run/current-system/sw/bin/bash`

### CODE REVIEW

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:48]: please use gpl3Only or gpl3Plus, this identifier is deprecated

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:14]:
```suggestion
  version = "unstable-2023-11-11";
```
any reason we can't use the stable version though?

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:25]: why is this needed?

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:30]: please add a comment explaining why

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:42]: why create it in marcel-start just to move it later?

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:49]: please set `meta.mainProgram` and `meta.platforms`

File: pkgs/shells/marcel/changes.patch
Comment (eclairevoyant) [inline pkgs/shells/marcel/changes.patch:10]: this requires bash to be on path, this seems incorrect. if anything, `substituteInPlace` fits better than using a patch.

File: pkgs/shells/marcel/default.nix
Comment (eclairevoyant) [inline pkgs/shells/marcel/default.nix:36]: shouldn't assume that `/usr/bin/env` exists nor that `sh` is on path

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/shells/marcel/default.nix
Comment (arcuru) [reply pkgs/shells/marcel/default.nix:30]: @KUD-00 commented in the thread in the marcel repo saying that the tests write to the HOME directory: https://github.com/geophile/marcel/issues/16#issuecomment-1810423637

Maybe this would allow them to run?
```nix
preCheck = ''
    export HOME=$(mktemp -d)
'';
```

File: pkgs/shells/marcel/default.nix
Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: Using stable version now.

File: pkgs/shells/marcel/default.nix
Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: my mistake. it's useless.

File: pkgs/shells/marcel/default.nix
Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: my mistake. can write over the generated file.

File: pkgs/shells/marcel/changes.patch
Comment (KUD-00) [reply pkgs/shells/marcel/changes.patch]: now using substituteInPlace and ${bash}/bin/bash, will you check for me whether i do it right or not?

File: pkgs/shells/marcel/default.nix
Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: I tried `export HOME=$(mktemp -d)`, and it partially solved the problem, but there're some test like this:

```bash
ls -0 ~root | map (f: f.path) failed:
    expected:
<<<['/root']>>>
    actual:
<<<['/build']>>>
```

many tests based on hardcoded '/tmp' folder, and many of them use `sudo`, i think these won't work with nix sandbox environment.

@eclairevoyant  @arcuru What do you think? Or is it better to `doCheck = false`, I have already add a comment to this line.

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:46]:
```suggestion
  postFixup = ''
    substituteInPlace $out/bin/marcel \
        --replace python3 ${python3}/bin/python3
    wrapProgram $out/bin/marcel \
        --prefix PYTHONPATH : "$program_PYTHONPATH"
  '';
```
There's no need to manually recreate the `marcel` script, it is automatically done.

Just having `marcel`'s own `PYTHONPATH` is not enough, because there are dependencies.
`$program_PYTHONPATH` is a hacky way around the fact that this is not actually a python script.
`$program_PYTHONPATH` is set after `python` wraps all python scripts. This is why I put this into `postFixup` because that runs after that.

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:11]:
```suggestion
{ lib
, buildPythonPackage
, fetchFromGitHub
, dill
, psutil
, python3
, bash
}:
```
There are some unused imports.
(`makeWrapper` is provided by `buildPythonPackage`)

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:13]: It might be more appropriate to use [buildPythonApplication](https://ryantm.github.io/nixpkgs/languages-frameworks/python/#buildpythonapplication-function).
If you do this you'll have to do `callPackage` instead of `python3Packages.callPackage` in `all-packages.nix`.
Look at the linked page for more info.

Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: thank you! now change to use buildPythonApplication.

File: pkgs/shells/marcel/default.nix
Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: It works! Thank you!

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:2]:
```suggestion
```

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:10]: There has been a `v0.20.0` release yesterday, which fixed most of the hard-coded `/bin/bash` stuff, so that no longer needs patching.
Though, there hasn't been an actual GitHub release, only a tag. So until there's a release, you can use `rev="refs/tags/v${version}"` to fetch by tag.

(When changing versions, don't forget to change/delete the hash, else it will still use the previous version)
```suggestion
  version = "0.20.0";
```

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:38]:
```suggestion
  postFixup = ''
    substituteInPlace $out/bin/marcel \
        --replace python3 ${python3}/bin/python3
    wrapProgram $out/bin/marcel \
        --prefix PATH : "$program_PATH:${bash}/bin" \
        --prefix PYTHONPATH : "$program_PYTHONPATH"
    '';
```
It's recommended to use indentation (4 spaces) in multiline shell commands.

Also, please add `$program_PATH` and `${bash}/bin` to `$PATH`.
- `$program_PATH` is just `$out/bin`, but this is variable is what the original python wrapper uses, so I'll recommend this. The reason this is used is for it to be able to find `farcel.py` (for running sudo commands)
- `${bash}/bin` is needed in case some weird person somehow doesn't have bash installed. (The program assumes that `bash` is available)

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [inline pkgs/shells/marcel/default.nix:29]:
```suggestion
```

Can be removed when updating to `v0.20.0`

**Review by TomaSajt — CHANGES_REQUESTED** [review]
I requested some changes about using the latest version

**Review by TomaSajt — COMMENTED** [review]
LGTM
Please, could you restructure your commits, so that there are only 2 commits: one called `maintainers: add kud`, which only has the `maintainer-list.nix` change, and another one called `marcel: init at 0.20.0`, which has every other change.
If you need help with this, I can help.

**Comment by KUD-00** [comment]
@TomaSajt now it seems only contains 2 commits. Thank you for your instructions!

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:42072]: Use the new pkgs/by-name hierarchy.

File: pkgs/shells/marcel/default.nix
Comment (AndersonTorres) [inline pkgs/shells/marcel/default.nix:30]: Have you ever tested this?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/shells/marcel/default.nix
Comment (TomaSajt) [reply pkgs/shells/marcel/default.nix:30]: This was the original script:
https://github.com/geophile/marcel/blob/master/bin/marcel
it's just a `bash` script which calls `python3`

This is why the `--prefix PYTHONPATH : "$program_PYTHONPATH"` hack needed to be used:  the default python wrapping logic would never wrap bash scripts

File: pkgs/top-level/all-packages.nix
Comment (KUD-00) [reply pkgs/top-level/all-packages.nix]: now using new by-name hierarchy

File: pkgs/shells/marcel/default.nix
Comment (KUD-00) [reply pkgs/shells/marcel/default.nix]: I have tested marcel built by this script on my PC, and it works well. Any suggestions about this?

File: pkgs/by-name/ma/marcel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/ma/marcel/package.nix:30]: This is strange.
Usually wrapProgram is a script available from makeWrapper package, but you are using it without the package...

File: pkgs/by-name/ma/marcel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/ma/marcel/package.nix:41]:
```suggestion
```

It is already set by buildPythonApplication.

File: pkgs/by-name/ma/marcel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/ma/marcel/package.nix:25]:
```suggestion
  # The tests use sudo and try to read/write $HOME/.local/share/marcel and /tmp
  doCheck = false;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ma/marcel/package.nix
Comment (TomaSajt) [reply pkgs/by-name/ma/marcel/package.nix:30]: I believe that because normal python scripts need to be wrapped (for the pythonpath), makeWrapper is included when building a python package

File: pkgs/by-name/ma/marcel/package.nix
Comment (KUD-00) [reply pkgs/by-name/ma/marcel/package.nix]: it is removed now

**Review by AndersonTorres — APPROVED** [review]

**Comment by MaxwellDupre** [comment]
format = "setuptools";
is now
 pyproject = true;

And usually the these changes are split (ie 2 PRs):
1) for maintainers-list
2) marcel

**Comment by eclairevoyant** [comment]
> And usually the these changes are split (ie 2 PRs):
> 
>     1. for maintainers-list
> 
>     2. marcel

No, they are usually one PR, adding to maintainers list isn't usually approved unless you have some package to maintain.

**Comment by AndersonTorres** [comment]
> And usually the these changes are split (ie 2 PRs):

No need for a new Pull Request. A commit suffices

> adding to maintainers list isn't usually approved unless you have some package to maintain.

This is at the discretion of the committers. Sometimes the person wants to contribute to multiple packages at once, then it  is better to do a PR for maintainers list update first.

**Comment by KUD-00** [comment]
now using ` pyproject = true;`

I'm not sure why it seems I need to add setuptools into propagatedBuildInputs, according to this [comment](https://github.com/NixOS/nixpkgs/issues/253154#issuecomment-1712673789), `pyproject = true` should equals to `format = pyproject`.

File: pkgs/by-name/ma/marcel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/ma/marcel/package.nix:24]:
```suggestion
  propagatedBuildInputs = [
    bash
  ] ++ (with python3.pkgs; [
    dill
    psutil
    setuptools
  ]);
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ma/marcel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/ma/marcel/package.nix:32]: Nit: use only two spaces to indent.
```suggestion
      --replace python3 ${python3}/bin/python3
```

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/by-name/ma/marcel/package.nix
Comment (KUD-00) [reply pkgs/by-name/ma/marcel/package.nix:32]: https://github.com/NixOS/nixpkgs/pull/267461#discussion_r1396529889

should i use two space indention?

Comment (eclairevoyant) [reply pkgs/by-name/ma/marcel/package.nix:32]: yes, always 2 spaces in nix.

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
Thanks for the contribution and patience - one more point of feedback from my side before I'd consider it mergeable. Also FYI 0.22.2 is out

File: pkgs/by-name/ma/marcel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ma/marcel/package.nix:24]: setuptools should be in `nativeBuildInputs` not `propagatedBuildInputs` as it is just used to build the package but not required afterwards. that should be fixed, the rest of the comment is just nits.


bash can just be in `buildInputs` and can use pythonPath for the rest rather than propagatedBuildInputs. propagatedBuildInputs would be needed if you were importing marcel in some other python package, but since this is an application, you're not, so `pythonPath` is sufficient.

this whole thing could be written as:
```nix
  nativeBuildInputs = with python3Packages; [
    setuptools
  ];

  buildInputs = [
    bash
  ];

  pythonPath = with python3Packages; [
    dill
    psutil
  ];
```

File: pkgs/by-name/ma/marcel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ma/marcel/package.nix:32]:
```suggestion
```
this is not needed as python3 is added to the path of the wrapper.

File: pkgs/by-name/ma/marcel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ma/marcel/package.nix:34]:
```suggestion
      --prefix PATH : "$program_PATH:${lib.getBin bash}/bin" \
```

**Review by ghost — COMMENTED** [review]
looks good -- minor issue with setuptools, it should not get propagated as it is just used to build the package and the substituteInPlace is not needed as python3 is already in the path.

validated maintainer info matches:
```
curl -s https://api.github.com/users/KUD-00|jq '{name,github:.login,githubId:.id}'
```

**Review by ghost — APPROVED** [review]

**Pull request merged**
